src/ControllerMessenger.ts

```diff
@@ -385,13 +403,20 @@ export class ControllerMessenger<
     eventType: E,
     ...payload: ExtractEventPayload<Event, E>
   ) {
-    const subscribers = this.events.get(eventType) as Set<
-      ExtractEventHandler<Event, E>
-    >;
+    const subscribers = this.events.get(eventType);
 
     if (subscribers) {
-      for (const eventHandler of subscribers) {
-        eventHandler(...payload);
+      for (const [handler, selector] of subscribers.entries()) {
+        if (selector) {
+          const newValue = selector(...payload);
+
+          if (newValue !== this.eventPayloadCache.get(handler)) {
+            this.eventPayloadCache.set(handler, newValue);
+            handler(newValue);
+          }
+        } else {
+          handler(...payload);
+        }
```

Comment:
This is a pre-existing issue so I'm disinclined to deal with it here, but, if a handler or selector throws, any remaining handlers will not be called. Should we somehow catch errors from handlers and/or selectors?

cc: @Gudahtt

Reply:
I suppose we could just repeat the trick we use in `safe-event-emitter`:  https://github.com/MetaMask/safe-event-emitter/blob/faf5620ac17ac6d32d8dd00ba0c3ffd9247ee6cf/index.ts/#L8-L17

Reply:
Ref: #566